refactor(elixir): route from router.route/1 to worker.router/2 #3654

lordB8r opened the pull request:
## Current behavior

All messages routed through Router.route/1.
```elixir
Router.route(%Ockam.Message{payload: pl, onward_route: o_r, return_route: r_r})
``` 

## Proposed Changes

Move the function to Worker.route/2 and add a state field for future usage.

```elixir
Router.route(%Ockam.Message{payload: pl, onward_route: o_r, return_route: r_r}, state: state)
``` 

All functions calling Router.route/1 will now call Worker.route/2 and pass `state` or `nil`

Set default `state` or `nil` when state unknown.

This is in reference to feature request #3654 

## Checks

<!-- To help us review and merge this pull request quickly, please confirm the following:  -->

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/edit/main/CONTRIBUTORS.csv) file in the github web UI and create a PR, this will mark the commit as verified.

polvorin commented on examples/elixir/get_started/echoer.exs:
```diff
@@ -2,12 +2,12 @@ defmodule Echoer do
   def handle_message(message, %{address: address} = state) do
     IO.puts("Address: #{address}\t Received: #{inspect(message)}")
-    Router.route(Message.reply(message, address, Message.payload(message)))
+    Worker.route(Message.reply(message, address, Message.payload(message)), state)
```
This is a good example of where the change make sense.  This _is_ a worker,  and you have the worker state to pass,   `Worker.route(message, state)` make sense.   And that right now just discard the state parameter and call Router.route/1,  that's ok and what the issue was about 👍

polvorin commented on implementations/elixir/ockam/ockam/lib/ockam/examples/forwarding/forwarding.ex:
```diff
@@ -65,11 +65,14 @@ defmodule Ockam.Examples.Forwarding do
-      return_route: ["example_initiator"],
-      payload: "Hello forward"
-    })
+    Ockam.Worker.route(
```
this doesn't make sense,  it is _not_  a worker making this call, it doesn't have any state to pass around (as you noted passing it a nil) .   The relevant part of the issue is
> and changing all usages of Ockam.Router.route/1 from context of workers to Ockam.Worker.route/2

In general,   there shouldn't be any case in which Ockam.Worker.route/2 is called with _nil_ as state.

lordB8r replied:
Good point, though I inquired [about this ](https://github.com/build-trust/ockam/issues/3654#issuecomment-1441232891) and was under the impression that *all* instances would use Worker, not Route, but I can now understand the confusion and will revert where it's not necessary.

lordB8r replied:
@polvorin For clarification, in `examples/messaging/shuffle.ex`, the function originally was: 
```elixir
def forward_message(message, _state)
  ...
  Ockam.Router.route(Message.forward(message))
```
My PR updates it to make state available (because it was ignored), and then pass it into the Worker. Is this the desired pattern?

polvorin commented on implementations/elixir/ockam/ockam/test/ockam/secure_channel_test.exs:
```diff
@@ -389,14 +392,62 @@ defmodule Ockam.SecureChannel.Tests do
     }
   end
 
+  test "dynamic identity" do
```
this looks strange,  these tests where removed some time ago,  and should be failing now if they ran due to changes on the API.
So need to remove these,   probably they come from some conflicts during a merge/rebase on your branch?

lordB8r replied:
I am also not sure about why these are here if they were removed. I was keeping my branch up to date with Github's merge function. I'll review and remove these.

lordB8r replied:
I've removed the old test in commit https://github.com/build-trust/ockam/commit/a6e3faeece3e6c6f3d2dee05d9269e9ea9152151.

polvorin reviewed (changes requested):
Hi @lordB8r , thanks for your work!

Looks like you changed all occurences of route(),  while only those on a Worker should had been changed,  let a couple of comments to indicate where it make sense,  and where it need to be back to the original. 
Also there is some strangeness on test files,  let a comment there as well.

lordB8r committed:
feat(elixir): provide a route/2 function on worker module

Ockam.Worker.route/2 , that delegates to Ockam.Router.route/1,
the additional argument is the worker' state, unused at the moment.

lordB8r commented:
@polvorin @mrinalwadhwa Requested changes updated 2 weeks ago. Is this good now? Or need more changes? Please advise.

polvorin reviewed (approved):
Hi @lordB8r , thanks for your contribution! 👏 .

I force pushed some minor changes on your branch:
* Changed the commit message to adhere to our enforced standard
* There was a commented-out line on a test case, that was uncommented on your original PR, that caused the test to fail.  I commented it back (having that code commented out is bad in itself, we will remove that on a separate PR)
* Removed some trailing whitespace that our linters where complaining about

Once the test pass on CI, the PR should be 🟢  to merge. 

In order to merge your contributions to Ockam,  you need to have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/.github/blob/main/CLA.md).  Once you have read the Agreement please open a new PR and add your Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam/blob/develop/.github/CONTRIBUTORS.csv) (that's the only change to be included in the PR). 

The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam/edit/develop/.github/CONTRIBUTORS.csv) file in the github web UI and create a separate Pull Request.    You can read more about the process on the [CONTRIBUTING readme](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#pull-requests) for
